### PR TITLE
MySQL Message Queue

### PR DESCRIPTION
**Pull request by pauljamescleary**
Create a MySQL Message Queue implementation.

* Created a `MySqlMessageQueue` in the mysql sub module
* Created a `MySqlMessage` that implements `CommandMessage` from core
* Created a `MessageType` enum to determine which type of command is on the message
* Created a `MySqlMessageQueueIntegrationSpec` which exercises 100% code coverage on the queue, including hidden behavior in the SQL

### CODE REVIEW

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+  private val logger = LoggerFactory.getLogger(classOf[MySQLMessageQueue])
+
+  private val INSERT_MESSAGE =
+    sql"""
+      |INSERT INTO message_queue(id, message_type, in_flight, data, created, updated, timeout_seconds, attempts)
+      |     VALUES ({id}, {messageType}, {inFlight}, {data}, {created}, {updated}, {timeoutSeconds}, {attempts})
+      |ON DUPLICATE KEY UPDATE updated={updated}
+    """.stripMargin
+
```
Comment (pauljamescleary) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: This is important for error handling.

The `ON DUPLICATE KEY` makes this idempotent.  We could in theory use `REPLACE INTO`, but that does a `DELETE` then an `INSERT`, whereas `ON DUPLICATE KEY` does an update.

Should not do `INSERT IGNORE INTO` as that will cause **all** errors (even valid SQL errors) to be ignored.  We do not want that.

In a nutshell, this enforces an "all or nothing" approach to sending messages.  If there are valid SQL errors, the entire batch of messages will fail.  If there are duplicate keys (for idempotentcy), but no SQL errors, the entire batch will **succeed**

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+  private val REQUEUE_MESSAGE =
+    sql"""
+      |UPDATE message_queue
+      |   SET in_flight = 0, updated = NOW()
+      | WHERE id = ?
+    """.stripMargin
+
```
Comment (pauljamescleary) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: `Requeue` clears the `in_flight` flag, allowing this message to be picked up on a different receive.

Note: we do not update the `attempts` here, that is updated when we actually claim the messages

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+  private val FETCH_UNCLAIMED =
+    sql"""
+         |SELECT id, message_type, data, attempts, timeout_seconds
+         |  FROM message_queue
+         | WHERE in_flight = 0
+         |    OR updated < DATE_SUB(NOW(),INTERVAL timeout_seconds SECOND)
+         | LIMIT ?
+    """.stripMargin
```
Comment (pauljamescleary) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: fetching unclaimed is interesting, so follow the logic here:

1. If the in_flight status is 0, we can take it
1. If in_flight is 1, however the the message visibility has expired, we can claim it.

That second statement is like SQS today.  In SQS, in theory 2 separate processes can process the same message at the same time.  For example, if the message timeout was 5 seconds, and a process had it queued but did not process it for say 10 seconds without changing the timeout OR requeue.  Large GC pauses can obviously force this to happen.

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+  // Important!  Make sure to update the updated timestamp to current, and increment the attempts!
+  private val CLAIM_MESSAGES =
+    sql"""
+         |UPDATE message_queue
+         |   SET in_flight=1, updated=NOW(), attempts=attempts+1
+         | WHERE id in (?)
+    """.stripMargin
+
```
Comment (pauljamescleary) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: when we claim a message, we make sure to mark the in flight status so noone else will pull it.

We also set the `updated` timestamp to reset message visibility

Finally, we update the number of attempts so we do not retry indefinitely on a message that maybe malformed and will just never succeed.

File: modules/core/src/main/scala/vinyldns/core/queue/MessageQueue.scala [hunk 5]
```diff
@@ -22,12 +22,9 @@ import vinyldns.core.domain.zone.ZoneCommand
...
 // $COVERAGE-OFF$
-
-// Message handle is implementation specific.  For example, in SQS, this may be the `Message` itself
-trait MessageHandle
-
-// Represents a command encoded in a message on a queue
-final case class CommandMessage(handle: MessageHandle, command: ZoneCommand)
+trait CommandMessage {
+  def command: ZoneCommand
+}
```
Comment (pauljamescleary) [inline modules/core/src/main/scala/vinyldns/core/queue/MessageQueue.scala]: Subtle change in the interface.  Since we do not use `MessageHandle` anywhere in any of the function signatures, it is confusing and unnecessary.  Just use the `CommandMessage` trait instead.  It makes message queue implementations simpler.

File: modules/api/src/test/scala/vinyldns/api/backend/CommandHandlerSpec.scala [hunk 6]
```diff
@@ -32,7 +32,7 @@ import vinyldns.api.domain.dns.DnsConnection
 import vinyldns.core.domain.batch.BatchChangeRepository
 import vinyldns.core.domain.record.{RecordChangeRepository, RecordSetChange, RecordSetRepository}
 import vinyldns.core.domain.zone._
-import vinyldns.core.queue.{CommandMessage, MessageCount, MessageHandle, MessageQueue}
+import vinyldns.core.queue.{CommandMessage, MessageCount, MessageQueue}
```
Comment (pauljamescleary) [inline modules/api/src/test/scala/vinyldns/api/backend/CommandHandlerSpec.scala]: changes in this file are needed to accomodate the small refactor of the `CommandMessage` type

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+/* MySQL Command Message implementation */
+final case class MySQLMessage(
+    id: MessageId,
+    attempts: Int,
+    timeout: FiniteDuration,
+    command: ZoneCommand)
+    extends CommandMessage
+object MySQLMessage {
+  final case class UnsupportedCommandMessage(msg: String) extends Throwable(msg)
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: this error can probably live in core, i imagine the other queues will have a similar casting structure

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+  def fromInt(i: Int): Either[InvalidMessageType, MessageType] = i match {
+    case 1 => Right(RecordChangeMessageType)
+    case 2 => Right(ZoneChangeMessageType)
+    case _ => Left(InvalidMessageType(i))
+  }
+}
+
+/* MySQL Command Message implementation */
+final case class MySQLMessage(
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: split MySQLMessage into its own file I think. I feel like we have a lot shoved in here

File: modules/mysql/src/main/resources/db/migration/V3.1__MessageQueue.sql [hunk 9]
```diff
@@ -0,0 +1,21 @@
+CREATE SCHEMA IF NOT EXISTS ${dbName};
```
Comment (rebstar6) [inline modules/mysql/src/main/resources/db/migration/V3.1__MessageQueue.sql]: just wondering, any reason we did 3.1 vs 4 here?

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+import vinyldns.proto.VinylDNSProto
+
+import scala.concurrent.duration._
+
+object MySQLMessageQueue {
+  final case class InvalidMessageHandle(msg: String) extends Throwable(msg)
+  final case class MessageAttemptsExceeded(msg: String) extends Throwable(msg)
+  final case class InvalidMessageTimeout(timeout: Int)
+      extends Throwable(s"Invalid message timeout $timeout")
+  final case class MessageId(value: String) extends AnyVal
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: that's the AnyVal extension give us?

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 2 above)
Comment (rebstar6) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: not certain that we need to change updated here, but i guess it doesnt matter either way

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+  private val CHANGE_TIMEOUT =
+    sql"""
+      |UPDATE message_queue
+      |   SET timeout_seconds = ?
+      | WHERE id = ?
+    """.stripMargin
+
+  private val FETCH_UNCLAIMED =
+    sql"""
+         |SELECT id, message_type, data, attempts, timeout_seconds
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: indenting is inconsistent from here down (with the 1st 3 sqls)

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+    *
+    * 2. If there are any errors stemming from step 1, delete them
+    *
+    * 3. For all the good messages, mark them as in-flight, and increment the attempts
+    */
+  def receive(count: MessageCount): IO[List[CommandMessage]] =
+    monitor("queue.JDBC.receive") {
+      IO {
+        // Need a max count so the user doesn't try to take 1MM messages, 10 is sufficient
+        val limit = Math.min(10, count.value)
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: 10 feels low, no?

Also, if were gonna say MessageCount has a max, AND we already have a positive restriction on the type, why not just build that into the type?

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+          // Mark the ones we got successfully
+          claimMessages(successes.map(_.id))
+          successes
+        }
+      }
+    }
+
+  def requeue(message: CommandMessage): IO[Unit] =
+    monitor("queue.JDBC.requeue") {
+      IO.fromEither(MySQLMessage.cast(message)).map { mysql =>
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: do we need all this casting everywhere, or should we just store ID on the MessageQueue type? Looks like SQS Message type also has a `private String messageId`

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+          }
+        }
+        .as(())
+    }
+
+  def send[A <: ZoneCommand](messages: NonEmptyList[A]): IO[SendBatchResult] =
+    monitor("queue.JDBC.sendBatch") {
+      IO {
+        DB.localTx { implicit s =>
+          // Note, we do replace into, so these really cannot fail, they all succeed or all fail
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: we do INSERT INTO?

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+  def insertParams(cmd: ZoneCommand): Seq[(Symbol, Any)] = {
+    val ts = DateTime.now
+    Seq(
+      'id -> cmd.id,
+      'messageType -> MessageType.fromCommand(cmd).value,
+      'inFlight -> 0,
+      'attempts -> 0,
+      'data -> getBytes(cmd),
+      'created -> ts,
+      'updated -> ts,
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: feels like updated should be none to start, no?

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 1 above)
Comment (rebstar6) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: interesting. updated does kinda feel like a weird field to be changing here though. in general, i think this will all work having reviewed the whole file, but some of the treatment of the updated field in general seems a little confusing

Comment (rebstar6) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: more clearly, treating updated strictly as 'time last pulled from queue' is what we want - by messing with it on creation, im worried we're complicating things

File: modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+  attempts: Int)
+
+final case class InvalidMessage(command: ZoneCommand) extends CommandMessage
+
+class MySQLMessageQueueIntegrationSpec extends WordSpec with Matchers
+  with BeforeAndAfterEach with EitherMatchers with EitherValues with BeforeAndAfterAll with ProtobufConversions {
+
+  private val underTest = new MySQLMessageQueue()
+
+  private val okZone: Zone = Zone("ok.zone.recordsets.", "test@test.com", adminGroupId = "group")
```
Comment (rebstar6) [inline modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala]: fyi, you can import:

vinyldns.core.TestZoneData.okZone
vinyldns.core.TestRecordSetData.aaaa

private val rsChange = vinyldns.core.TestRecordSetData.makeTestAddChange(aaaa, okZone)
private val zoneChange = vinyldns.core.TestZoneData.zoneChangePending (slightly different change, but still should work)

File: modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+      insert(rsChange.id, RecordChangeMessageType.value, true, rsChangeBytes, initialTs, initialTs, 30, 1)
+      val msgs = underTest.receive(MessageCount(1).right.value).unsafeRunSync()
+      msgs should have length 1
+    }
+  }
+
+  "changeMessageTimeout" should {
+    "update the message time out" in {
+      underTest.send(rsChange).unsafeRunSync()
+      underTest.changeMessageTimeout(testMessage, 100.seconds).unsafeRunSync()
```
Comment (rebstar6) [inline modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala]: dont like sleeping in a test, but is it worth changing one to 1 second and testing that it gets pulled a 2nd time?

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 1 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: I modified to update created  Problem is we have to update something ON DUPLICATE KEY.  Better than updated, less confusing I suppose

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 2 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: I pulled it out, and updated the integration spec

File: modules/mysql/src/main/resources/db/migration/V3.1__MessageQueue.sql (hunk 9 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/resources/db/migration/V3.1__MessageQueue.sql]: grr, we should not have been updating "major" version in flyway, realistically we should be on V1.4

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 10 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: probably don't need it.

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 11 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: yea, that is annoying, will fix assuming that scalafmt doesn't "fix it back"

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 12 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: `MessageCount` is in core, I suppose we could put that restriction in core.

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 13 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: Yea, I thought the same thing too.  I just don't know what other commands may come about.  I would be ok with putting it on the `CommandMessage` though, that would simplify things.  Right now it is really annoying to do this casting

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 14 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: Yep, INSERT INTO with duplicate key handling

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 15 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: updated really should be "NOW", I may modify that INSERT statement

File: modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala (hunk 16 above)
Comment (pauljamescleary) [reply modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala]: Ahhhh, `TestZoneData`!  I saw `VinylDNSTestData` and it was in the API.  Ok, will do

File: modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala (hunk 17 above)
Comment (pauljamescleary) [reply modules/mysql/src/it/scala/vinyldns/mysql/repository/MySQLMessageQueueIntegrationSpec.scala]: that's a good idea

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala (hunk 14 above)
Comment (rebstar6) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: comment says `replace into`

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+      IO {
+        // Need a max count so the user doesn't try to take 1MM messages, 10 is sufficient
+        val limit = Math.min(10, count.value)
+
+        DB.localTx { implicit s =>
+          // get unclaimed messages, note these could fail during retrieval
+          val claimed = fetchUnclaimed(limit)
+
+          // Errors could not be deserialized, have an invalid type, or exceeded retries
+          val errors = claimed.collect {
```
Comment (mitruly) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySQLMessageQueue.scala]: Was playing around with this a bit, and `claimed.separate` might make things easier to work with since it will separate a collection of `Either` into `Left`s and `Right`s.

File: modules/mysql/src/it/scala/vinyldns/mysql/queue/MySqlMessageQueueIntegrationSpec.scala [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+      val r = underTest.receive(MessageCount(1).right.value).unsafeRunSync()
+
+      // second receive should not re-deliver
+      underTest.receive(MessageCount(1).right.value).unsafeRunSync() shouldBe empty
+
+      underTest.changeMessageTimeout(testMessage, 1.seconds).unsafeRunSync()
+      Thread.sleep(2000)
+      val r2 = underTest.receive(MessageCount(1).right.value).unsafeRunSync()
+      r2 should have length 1
+      r.headOption shouldBe r2.headOption`
```
Comment (mitruly) [inline modules/mysql/src/it/scala/vinyldns/mysql/queue/MySqlMessageQueueIntegrationSpec.scala]: A wild ` appeared!

Comment (rebstar6) [reply modules/mysql/src/it/scala/vinyldns/mysql/queue/MySqlMessageQueueIntegrationSpec.scala]: CATCH IT!!!

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySqlMessageQueue.scala [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+      IO {
+        // Need a max count so the user doesn't try to take 1MM messages, 10 is sufficient
+        val limit = Math.min(10, count.value)
+
+        DB.localTx { implicit s =>
+          // get unclaimed messages, note these could fail during retrieval
+          val claimed = fetchUnclaimed(limit)
+
+          // Errors could not be deserialized, have an invalid type, or exceeded retries
+          val errors = claimed.collect {
```
Comment (mitruly) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySqlMessageQueue.scala]: Was playing around with this a bit, and `claimed.separate` might make things easier to work with since it will separate a collection of `Either` into `Left`s and `Right`s.

Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySqlMessageQueue.scala]: It is a good idea, but then I still need to pull out the id, doing something unsavory like `.map(_._.2)`.  Going to keep it like this, but thanks for that info!

File: modules/mysql/src/main/scala/vinyldns/mysql/queue/MySqlMessageQueue.scala [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+    *
+    * 2. If there are any errors stemming from step 1, delete them
+    *
+    * 3. For all the good messages, mark them as in-flight, and increment the attempts
+    */
+  def receive(count: MessageCount): IO[List[CommandMessage]] =
+    monitor("queue.JDBC.receive") {
+      IO {
+        // Need a max count so the user doesn't try to take 1MM messages, 10 is sufficient
+        val limit = Math.min(10, count.value)
```
Comment (rebstar6) [inline modules/mysql/src/main/scala/vinyldns/mysql/queue/MySqlMessageQueue.scala]: leaving this here? im still team embed this in the MessageCount type, but wont hold it up for that

Comment (pauljamescleary) [reply modules/mysql/src/main/scala/vinyldns/mysql/queue/MySqlMessageQueue.scala]: The "max" seems to be queue specific.  There is a limit of 10 in SQS, but could be 100, 200, 1000 in other cases like RabbitMQ or redismq.  Even here we could receive more than 10 at a time.  Truth is, I am unclear of what the limit should be, 10 will always be fine, but 1,000,000 maybe too many.  Would need to do some benchmarking to figure out the appropriate limit.  Even then, the limit could be different based on your MySql setup.

We could loop until we satisfy the count as an alternative, but we may never return if there is a long gap (hours, days) between messages.  So that is not a good option.

Another alternative would be to encode the error in the return type, something like "MaxMessageCountExceeded".  `EitherT[IO, MaxMessageCountExceeded, List[CommandMessage]]`.  It certainly makes working with the queue more complicated.

**tl;dr** I think leaving this in is an ok tradeoff for simplicity purposes.
